extmod/modasyncio.c

```diff
@@ -179,6 +185,130 @@ STATIC mp_obj_t task_done(mp_obj_t self_in) {
 }
 STATIC MP_DEFINE_CONST_FUN_OBJ_1(task_done_obj, task_done);
 
+STATIC mp_obj_t task_add_done_callback(mp_obj_t self_in, mp_obj_t callback) {
+    assert(mp_obj_is_callable(callback));
+    mp_obj_task_t *self = MP_OBJ_TO_PTR(self_in);
+
+    if (TASK_IS_DONE(self)) {
+        // In CPython the callbacks are not immediately called and are instead
+        // called by the event loop.  However, MicroPython's event loop doesn't
+        // support `call_soon` to handle callback processing.
+        //
+        // Because of this, it's close enough to call the callback immediately.
+
+        // TODO: Omit second parameter if callback only accepts 1 to match CPython
+
+        mp_call_function_2(callback, self_in, self->data);
```

Comment:
Is it possible to determine how many parameters we can pass, or otherwise count the second parameter as something that can be ignored if the called function doesn't accept it?

Reply:
Actually, I'll just remove this comment for now -- I'd need to update this in a few places outside this PR so probably best as its own distinct change.

Reply:
FWIW I've seen that kind of callback logic done in pure python by just trying to call with both are in a try/except and retry with just one arg.
You can do the same thing in C, not sure if there's a clearer way.

Reply:
Good idea!  I'll follow up with that on another PR either way, going to keep the change isolated here unless it's something encouraged to get this PR mergable.